agent/xds/listeners.go

```diff
@@ -512,7 +548,87 @@ func (s *Server) makeUpstreamListenerIgnoreDiscoveryChain(
+	l := makeListener(name, addr, port)
+
+	tlsInspector, err := makeTLSInspectorListenerFilter()
+	if err != nil {
+		return nil, err
+	}
+	l.ListenerFilters = []envoylistener.ListenerFilter{tlsInspector}
+
+	// Make a FilterChain for each linked service
+	// Match on the cluster name,
+	for svc, _ := range cfgSnap.TerminatingGateway.ServiceGroups {
+		clusterName := connect.ServiceSNI(svc.ID, "", svc.NamespaceOrDefault(), cfgSnap.Datacenter, cfgSnap.Roots.TrustDomain)
+		resolver, hasResolver := cfgSnap.TerminatingGateway.ServiceResolvers[svc]
+
+		clusterChain, err := s.sniFilterChainTerminatingGateway(name, clusterName, svc, cfgSnap)
+		if err != nil {
+			return nil, fmt.Errorf("failed to make filter chain for cluster %q: %v", clusterName, err)
+		}
+		l.FilterChains = append(l.FilterChains, clusterChain)
+
+		// if there is a service-resolver for this service then also setup subset filter chains for it
+		if hasResolver {
+			// generate 1 filter chain for each service subset
+			for subsetName := range resolver.Subsets {
+				clusterName := connect.ServiceSNI(svc.ID, subsetName, svc.NamespaceOrDefault(), cfgSnap.Datacenter, cfgSnap.Roots.TrustDomain)
+
+				clusterChain, err := s.sniFilterChainTerminatingGateway(name, clusterName, svc, cfgSnap)
+				if err != nil {
+					return nil, fmt.Errorf("failed to make filter chain for cluster %q: %v", clusterName, err)
+				}
+				l.FilterChains = append(l.FilterChains, clusterChain)
+			}
+		}
+	}
+
+	err = injectConnectFilters(cfgSnap, token, l, false)
+
+	// This fallback catch-all filter ensures a listener will be present for health checks to pass
+	// Envoy will reset these connections since known endpoints are caught by filter chain matches above
+	tcpProxy, err := makeTCPProxyFilter(name, "", "terminating_gateway_")
+	if err != nil {
+		return nil, err
+	}
+	fallback := envoylistener.FilterChain{
+		Filters: []envoylistener.Filter{
+			{Name: "envoy.filters.network.sni_cluster"},
+			tcpProxy,
+		},
+	}
+	l.FilterChains = append(l.FilterChains, fallback)
+
+	return l, nil
+}
+
+func (s *Server) sniFilterChainTerminatingGateway(listener, cluster string, service structs.ServiceID, cfgSnap *proxycfg.ConfigSnapshot) (envoylistener.FilterChain, error) {
+	sniCluster, err := makeSNIClusterFilter()
+	if err != nil {
+		return envoylistener.FilterChain{}, err
+	}
+
+	// The cluster name here doesn't matter as the sni_cluster filter will fill it in for us.
+	tcpProxy, err := makeTCPProxyFilter(listener, "", fmt.Sprintf("terminating_gateway_%s_", service.String()))
+	if err != nil {
+		return envoylistener.FilterChain{}, err
+	}
+
+	return envoylistener.FilterChain{
+		FilterChainMatch: makeSNIFilterChainMatch(cluster),
+		Filters: []envoylistener.Filter{
+			sniCluster,
+			tcpProxy,
+		},
+		TlsContext: &envoyauth.DownstreamTlsContext{
+			CommonTlsContext:         makeCommonTLSContext(cfgSnap, cfgSnap.TerminatingGateway.ServiceLeaves[service]),
```

Comment:
`makeCommonTLSContext ` will probably panic if the leaf isn't present (due to ACLs or some other reason), since it's a pointer.

Should probably just skip creating the filterchain if that's the case. But what's a good way to surface this issue, rather than silently not creating a listener for an expected destination?